### PR TITLE
Optimize Gemma4 H200 MoE and extend attention

### PR DESCRIPTION
**Pull request by BBuf**
## Summary

- Add H200 Triton fused MoE configs for Gemma4 `E=128,N=704` normal and down projections.
- Tune Hopper extend-attention block sizes for `Lq=129..256` to reduce TTFT/TPOT on Gemma4 prefill-heavy serving.
- Add a small-batch (M ≤ 256) per-head Triton kernel for Gemma4 QKV RMSNorm to lower MTP draft latency.
- ~~Add a Gemma4-specific Triton routing kernel that fuses top-k selection, top-k softmax, and per-expert scale into `gemma4_topk_softmax_scale`.~~ Reverted (commit `27cb94c45`) due to BF16 precision regression on MoE routing — see "Reverted optimizations" below.
- ~~Replace `Gemma4Router.norm(x) * fused_scale` with a single fused `rmsnorm(x, fused_scale, eps)` call on the CUDA fast path.~~ Reverted (commit `0d3c9c694`) for the same reason.
- Keep the optimization scoped to Gemma4/Hopper paths so existing model families keep their current behavior.

## Performance

Re-measured on `ion8-h200`, `sglang_bbuf`, single NVIDIA H200, TP=1, `google/gemma-4-26B-A4B-it`, BF16, `--attention-backend triton`, random `1024/256`, request rate 8, max concurrency 64, 80 prompts. Baseline = latest `origin/main` (`376635c1e`); patched = PR head (`27cb94c45`, with c2/c3 reverted).

| Workload | Build | Output tok/s | p50 TTFT ms | p50 TPOT ms | Median E2E ms | Result |
| --- | --- | ---: | ---: | ---: | ---: | --- |
| Random `1024/256`, 80 prompts | SGLang baseline (`376635c1e`) | 1670.72 | 96.56 | 20.30 | 5295.91 | Baseline |
| Random `1024/256`, 80 prompts | SGLang patched (`27cb94c45`) | **1725.50** | **74.04** | **16.62** | **4338.60** | **+3.28% output tok/s, -23.3% TTFT, -18.2% TPOT, -18.1% E2E** |

Full benchmark details:

| Build | Completed | Duration s | Request throughput req/s | Total tok/s | Median E2E ms |
| --- | ---: | ---: | ---: | ---: | ---: |
| SGLang baseline (`376635c1e`) | 80 | 12.26 | 6.526 | 8353.59 | 5295.91 |
| SGLang patched (`27cb94c45`) | 80 | 11.87 | 6.740 | 8627.50 | 4338.60 |

The remaining speedup comes from the H200 MoE Triton configs + Hopper extend-attention block-size tuning + small-batch QKV RMSNorm; the heavier routing-fusion optimizations were dropped to keep accuracy on the MTP path.

## Accuracy

Validated on `ion8-h200` via the registered Gemma4 MTP CI test (`test_gemma4_mtp_26b_a4b_extra::test_gsm8k_mtp`, TP=2, `--enable-deterministic-inference`, NEXTN spec decode, 200 GSM8K examples, 5-shot). With the two reverted commits, GSM8K MTP score is back to baseline:

| Build (TP=2, deterministic) | GSM8K MTP topk=1 | avg accept length | Threshold 0.41 |
| --- | ---: | ---: | --- |
| SGLang baseline (`376635c1e`, all reverts = main) | 0.445 | 4.494 | ✓ pass |
| SGLang patched (`27cb94c45`, c1+c4 kept, c2+c3 reverted) | 0.445 | 4.494 | ✓ pass |
| ~~PR head before revert (`10ab189e3`, all 4 commits)~~ | ~~0.360~~ | ~~4.475~~ | ✗ fail (-5pp) |

Standalone GSM8K (TP=1, no MTP, 8-shot, 200 questions, no deterministic mode) shows the typical run-to-run variance (~5pp swing) but no systematic regression — patched and baseline land within noise of each other (patched: 0.41 / 0.495 across two runs; baseline: 0.465 / 0.445).

## Reverted optimizations

Bisecting `test_gemma4_mtp_26b_a4b_extra` on `ion8-h200` (TP=2, deterministic):

| Configuration | GSM8K topk=1 | Δ vs PR head |
| --- | ---: | ---: |
| PR head (all 4 commits) | 0.360 | 0 |
| Revert only `fabb5b5ee` (QKV RMSNorm small batch) | 0.360 | 0 — innocuous |
| Revert only `d72d246a3` (Router RMSNorm fuse) | 0.395 | +0.035 |
| Revert only `03826cdd9` (MoE topk fuse) | 0.380 | +0.020 |
| Revert only `3398cb7af` (extend-attn + MoE configs) | 0.360 | 0 — innocuous |
| **Revert `d72d246a3` + `03826cdd9` together** | **0.445** | **+0.085 (pass)** |
| Revert all 4 (= origin/main) | 0.445 | identical to c2+c3 revert |

Both reverted commits modify the Gemma4 MoE routing path in `gemma4_causal.py`:
- `d72d246a3`: `rmsnorm(x, fused_scale, eps)` fast path in `Gemma4Router.forward`. The math is equivalent to `(self.norm(x) with weight=1) * fused_scale`, but BF16 accumulation order in the fused kernel diverges from the two-step path when `fused_scale ≈ hidden_size**-0.5 ≈ 0.022`.
- `03826cdd9`: `_gemma4_topk_softmax_scale_kernel` doing topk + stable-softmax + per-expert scale in one pass. Equivalent to `softmax(topk_logits) * scale[topk_ids]` for `topk ≤ 8`, but its FP32-internal `exp(top_logit - top1) / sum_top_exp` ordering doesn't match `torch.nn.functional.softmax` bit-for-bit.

Either alone causes a ~0.02-0.035 GSM8K drop; together the routing logits drift enough to flip enough expert selections that the MTP draft/target distribution gap widens (MTP `avg_accept_length` still looks healthy at 4.48, masking the underlying quality regression).

If you want to recover these speedups, the kernels need to be made bit-equivalent to the eager paths — e.g. accept `weight=ones` in the fused router RMSNorm and apply `fused_scale` as a separate elementwise op, and have `_gemma4_topk_softmax_scale_kernel` mirror PyTorch's exact softmax floating-point order (compute max, subtract, exp, sum, divide — all in fp32 with the same reduction tree).

## Validation

- Bisect, perf re-measurement, and Gemma4 MTP accuracy validation all completed on `ion8-h200`, container `sglang_bbuf`.
- Python syntax check passed for the changed Python files.
- New H200 Triton config JSON files (`E=128,N=704`) parse successfully.
- Extend-attention block-size sanity checked for `Lq=128,192,256,288`.
- Server benchmark + Gemma4 MTP CI test (TP=2, deterministic, 200-question GSM8K) confirm patched ≡ baseline accuracy (both 0.445).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26944892458](https://github.com/sgl-project/sglang/actions/runs/26944892458)
Latest PR Test (Extra): :x: [Run #26944892240](https://github.com/sgl-project/sglang/actions/runs/26944892240)

### CODE REVIEW

**Comment by pyc96** [comment]
@BBuf Could you please take a look at https://github.com/sgl-project/sglang/pull/26502 for fused router and https://github.com/sgl-project/sglang/pull/25461 for norm ? We can potentially merge the efforts

Just quickly checked myself, https://github.com/sgl-project/sglang/pull/26502 seems better due to less top-k reduce complexity. https://github.com/sgl-project/sglang/pull/25461 seems complementary to your changes

**Comment by BBuf** [comment]
> @BBuf Could you please take a look at #26502 for fused router and #25461 for norm ? We can potentially merge the efforts
> 
> Just quickly checked myself, #26502 seems better due to less top-k reduce complexity. #25461 seems complementary to your changes

Ok, I'll review it.

File: python/sglang/srt/layers/gemma4_fused_ops.py [hunk 1]
```diff
@@ -191,6 +191,52 @@ def _gemma_qkv_rmsnorm_kernel(
             tl.store(V_ptr + off, out.to(V_ptr.dtype.element_ty), mask=mask)
 
 
+@triton.jit
+def _gemma_qkv_rmsnorm_by_head_kernel(
```
Comment (yuan-luo) [inline python/sglang/srt/layers/gemma4_fused_ops.py]: `_gemma_qkv_rmsnorm_by_head_kernel` (grid `(M, total_heads)`) and the existing `_gemma_qkv_rmsnorm_kernel` (grid `[(M,)]`, from #24696) compute the *same* per-head RMSNorm, the only difference is the parallelization (2D grid vs per-token loop). That's two kernels to keep in sync forever.

Can we collapse to one kernel?

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py [hunk 2]
```diff
@@ -82,8 +82,10 @@ def _get_block_sizes_for_extend_attention(Lq: int, Lv: int):
                 BLOCK_M, BLOCK_N = (16, 64)
         elif _is_cuda and CUDA_CAPABILITY[0] >= 9:
             # Hopper architecture (H100, etc.)
-            if Lq <= 256:
+            if Lq <= 128:
                 BLOCK_M, BLOCK_N = (128, 64)
+            elif Lq <= 256:
```
Comment (yuan-luo) [inline python/sglang/srt/layers/attention/triton_ops/extend_attention.py:87]: These thresholds (128/256) are tuned on H200 + gemma-4-26B.
Ideally these become autotune keys rather than hand-thresholds.

File: python/sglang/srt/layers/moe/moe_runner/triton_utils/configs/triton_3_6_0/E=128,N=704,device_name=NVIDIA_H200.json [hunk 3]
```diff
@@ -0,0 +1,114 @@
+{
+    "1": {
```
Comment (yuan-luo) [inline python/sglang/srt/layers/moe/moe_runner/triton_utils/configs/triton_3_6_0/E=128,N=704,device_name=NVIDIA_H200.json:2]: Could be good if adding H100 and H20 configs.

File: python/sglang/srt/layers/attention/triton_ops/extend_attention.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/srt/layers/attention/triton_ops/extend_attention.py:87]: Agree. This helper is already architecture-heuristic based today, and these Hopper thresholds are backed by the H200 + Gemma4 measurements in this PR. Turning the block-size choice into an autotune key is the better long-term direction, but I would prefer to handle that as a follow-up instead of blocking this PR on a broader autotune refactor.

File: python/sglang/srt/layers/moe/moe_runner/triton_utils/configs/triton_3_6_0/E=128,N=704,device_name=NVIDIA_H200.json (hunk 3 above)
Comment (BBuf) [reply python/sglang/srt/layers/moe/moe_runner/triton_utils/configs/triton_3_6_0/E=128,N=704,device_name=NVIDIA_H200.json:2]: Agree it would be good to have H100/H20 configs too. I do not currently have H100/H20 access for this shape, so I do not want to copy the H200 config without tuning/benchmark data. I'll keep this PR scoped to the measured H200 configs and leave H100/H20 tuning as a follow-up when we can validate on those cards.

File: python/sglang/srt/layers/gemma4_fused_ops.py (hunk 1 above)
Comment (BBuf) [reply python/sglang/srt/layers/gemma4_fused_ops.py]: Good point. Fixed in `847ce14af3`: `_gemma_qkv_rmsnorm_by_head_kernel` is removed, and the existing `_gemma_qkv_rmsnorm_kernel` now supports both launch shapes through a `BY_HEAD` constexpr. The small-batch path still launches with grid `(M, total_heads)`, while the large-batch path still launches with grid `(M,)`, but they share the same helper/body now.

Validated on `ion8-h200` / H200, GPU 0 with BF16 QKV/Q-only correctness and CUDA-event microbench against the parent commit:
- max abs error stayed within BF16 tolerance (max observed `0.015625`)
- parent: `M128_qkv 13.84us`, `M512_qkv 13.72us`
- patched: `M128_qkv 13.60us`, `M512_qkv 13.82us`
So this removes the duplicate maintenance surface without a measurable regression in this microbench.

**Review by yuan-luo — APPROVED** [review]
Please make sure CI pass. LGTM.
